server/storage/storage_gc_test.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"math"
+	"time"
+
+	. "github.com/pingcap/check"
+	"github.com/pingcap/failpoint"
+	"github.com/tikv/pd/server/storage/endpoint"
+)
+
+var _ = Suite(&testStorageGCSuite{})
+
+type testStorageGCSuite struct {
+}
+
+func testGCSafePoints() ([]string, []uint64) {
+	spaceIDs := []string{
+		"keySpace1",
+		"keySpace2",
+		"keySpace3",
+		"keySpace4",
+		"keySpace5",
+	}
+	safePoints := []uint64{
+		0,
+		1,
+		4396,
+		23333333333,
+		math.MaxUint64,
+	}
+	return spaceIDs, safePoints
+}
+
+func testServiceSafePoints() ([]string, []*endpoint.ServiceSafePoint) {
+	spaceIDs := []string{
+		"keySpace1",
+		"keySpace1",
+		"keySpace1",
+		"keySpace2",
+		"keySpace2",
+		"keySpace2",
+		"keySpace3",
+		"keySpace3",
+		"keySpace3",
+	}
+	expireAt := time.Now().Add(100 * time.Second).Unix()
+	serviceSafePoints := []*endpoint.ServiceSafePoint{
+		{ServiceID: "service1", ExpiredAt: expireAt, SafePoint: 1},
+		{ServiceID: "service2", ExpiredAt: expireAt, SafePoint: 2},
+		{ServiceID: "service3", ExpiredAt: expireAt, SafePoint: 3},
+		{ServiceID: "service1", ExpiredAt: expireAt, SafePoint: 1},
+		{ServiceID: "service2", ExpiredAt: expireAt, SafePoint: 2},
+		{ServiceID: "service3", ExpiredAt: expireAt, SafePoint: 3},
+		{ServiceID: "service1", ExpiredAt: expireAt, SafePoint: 1},
+		{ServiceID: "service2", ExpiredAt: expireAt, SafePoint: 2},
+		{ServiceID: "service3", ExpiredAt: expireAt, SafePoint: 3},
+	}
+	return spaceIDs, serviceSafePoints
+}
+
+func (s *testStorageGCSuite) TestSaveLoadServiceSafePoint(c *C) {
+	storage := NewStorageWithMemoryBackend()
+	testSpaceID, testSafePoints := testServiceSafePoints()
+	for i := range testSpaceID {
+		c.Assert(storage.SaveServiceSafePoint(testSpaceID[i], testSafePoints[i]), IsNil)
+	}
+	for i := range testSpaceID {
+		loadedSafePoint, err := storage.LoadServiceSafePoint(testSpaceID[i], testSafePoints[i].ServiceID)
+		c.Assert(err, IsNil)
+		c.Assert(loadedSafePoint, DeepEquals, testSafePoints[i])
+	}
+}
+
+func (s *testStorageGCSuite) TestLoadMinServiceSafePoint(c *C) {
+	// enable a custom timeout to accommodate async deletion delay
+	c.Assert(failpoint.Enable("github.com/tikv/pd/server/storage/endpoint/customTimeout", "return(true)"), IsNil)
```

Comment:
Enable the safepoint when you need, How about  enable the above safepoint before L107? which means before calling the function `LoadMinServiceSafePoint`